pkg/kubelet/status/state/state_checkpoint_test.go

```diff
@@ -0,0 +1,175 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager"
+)
+
+func newTestStateCheckpoint(t *testing.T) *stateCheckpoint {
+	// create temp dir
+	testingDir, err := os.MkdirTemp("", "pod_resource_allocation_state_test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		if err := os.RemoveAll(testingDir); err != nil {
+			t.Fatal(err)
+		}
+	}()
+	cache := NewStateMemory()
+	checkpointManager, err := checkpointmanager.NewCheckpointManager(testingDir)
+	require.NoError(t, err, "failed to create checkpoint manager")
+	checkpointName := "pod_state_checkpoint"
+	sc := &stateCheckpoint{
+		cache:             cache,
+		checkpointManager: checkpointManager,
+		checkpointName:    checkpointName,
+	}
+	return sc
+}
+
+func Test_stateCheckpoint_storeState(t *testing.T) {
+	sc := newTestStateCheckpoint(t)
+	type args struct {
+		podResourceAllocation PodResourceAllocation
+	}
+
+	tests := []struct {
+		name string
+		args args
+	}{}
+	suffix := []string{"Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "n", "u", "m", "k", "M", "G", "T", "P", "E", ""}
+	factor := []string{"1", "0.1", "0.03", "10", "100", "512", "1000", "1024", "700", "10000"}
+	for _, fact := range factor {
+		for _, suf := range suffix {
+			if (suf == "E" || suf == "Ei") && (fact == "1000" || fact == "10000") {
+				// when fact is 1000 or 10000, suffix "E" or "Ei", the quantity value is overflow
+				// see detail https://github.com/kubernetes/apimachinery/blob/95b78024e3feada7739b40426690b4f287933fd8/pkg/api/resource/quantity.go#L301
+				continue
+			}
+			tests = append(tests, struct {
+				name string
+				args args
+			}{
+				name: fmt.Sprintf("resource - %s%s", fact, suf),
+				args: args{
+					podResourceAllocation: PodResourceAllocation{
+						"pod1": {
+							"container1": {
+								Requests: v1.ResourceList{
+									v1.ResourceCPU:    resource.MustParse(fmt.Sprintf("%s%s", fact, suf)),
+									v1.ResourceMemory: resource.MustParse(fmt.Sprintf("%s%s", fact, suf)),
+								},
+							},
+						},
+					},
+				},
+			})
+		}
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := sc.cache.ClearState()
+			require.NoError(t, err, "failed to clear state")
+
+			defer func() {
+				err = sc.checkpointManager.RemoveCheckpoint(sc.checkpointName)
+				require.NoError(t, err, "failed to remove checkpoint")
+			}()
+
+			err = sc.cache.SetPodResourceAllocation(tt.args.podResourceAllocation)
+			require.NoError(t, err, "failed to set pod resource allocation")
+
+			err = sc.storeState()
```

Comment:
Rather than all this mucking with the internal cache and state, I'd prefer to test the external interfaces here. In otherwords, something like this:

```go
testDir = getTestDir(t)
originalSC, err := NewStateCheckpoint(testDir, testCheckpoint)
require.NoError(t, err)

err = originalSC.SetPodResourceAllocation(tt.args.podResourceAllocation)
require.NoError(t, err)

actual := originalSC.GetPodResourceAllocation()
assert.Equal(t, tt.args.podResourceAllocation, actual)

// additional verification of originalSC.GetContainerResourceAllocation(...)

newSC := NewStateCheckpoint(testDir, testCheckpoint)
require.NoError(t, err)

actual = newSC.GetPodResourceAllocation()
assert.Equal(t, tt.args.podResourceAllocation, actual)

// aditional verification of newSC.GetContainerResourceAllocation(...)
```

Reply:
I updated the test function following the pattern above, but replaced the assert function by function below

```go
func verifyPodResourceAllocation(t *testing.T, expected, actual *PodResourceAllocation, msgAndArgs string) {
	for podUID, containerResourceList := range *expected {
		require.Equal(t, len(containerResourceList), len((*actual)[podUID]), msgAndArgs)
		for containerName, resourceList := range containerResourceList {
			for name, quantity := range resourceList.Requests {
				if !quantity.Equal((*actual)[podUID][containerName].Requests[name]) {
					t.Errorf(msgAndArgs)
				}
			}
		}
	}
}
```

Because reflect.DeepEqual won't call Quantity.Equal function internally, and `assert.Equal(t, tt.args.podResourceAllocation, actual)` will fail because of different formats of Quantity.